### ISSUE TITLE
No module on build with kernel 5.14.0

### ISSUE THREAD

**Issue by NiKiZe** [Lochnair/xt_tls#52]
When running `make` nothing in `src` gets compiled.

This kernel is not the latest, but before going down the path of testing different kernels... my question is what is a known good kernel?

Trying to build src gives the following:
```
/xt_tls/src # make -j16
make -C /lib/modules/5.14.0-gentoo/build SUBDIRS=/root/xt_tls/src modules
make[1]: Entering directory '/usr/src/linux-5.14.0-gentoo'
  DESCEND objtool
  CALL    scripts/atomic/check-atomics.sh
  CALL    scripts/checksyscalls.sh
make[1]: Leaving directory '/usr/src/linux-5.14.0-gentoo'
```

After some investigation of how module builds work I came up with a quick patch, however this is copy and paste from places on the net without actually understanding much of it ;)

```diff
diff --git a/src/Makefile b/src/Makefile
index 97e2ae5..6e5e323 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,5 +1,5 @@
-obj-m += xt_tls.o
-xt_tls-objs := xt_tls_main.o hostset.o
+obj-m := xt_tls.o
+xt_tls-y := xt_tls_main.o hostset.o
 KERNEL_VERSION := $(shell uname -r)
 IDIR := /lib/modules/$(KERNEL_VERSION)/kernel/net/netfilter/
 KDIR := /lib/modules/$(KERNEL_VERSION)/build
@@ -7,13 +7,13 @@ DESTDIR ?= /
 PWD := $(shell pwd)
 VERSION := $(shell git rev-parse HEAD 2>/dev/null)
 default:
-           $(MAKE) -C $(KDIR) SUBDIRS=$(PWD) modules
+           $(MAKE) -C $(KDIR) M=$(PWD) modules

 debug:
            $(MAKE) -C $(KDIR) SUBDIRS=$(PWD) CFLAGS_MODULE="-DXT_TLS_DEBUG" modules

 install:
-           $(MAKE) -C $(KDIR) SUBDIRS=$(PWD) INSTALL_MOD_PATH=$(DESTDIR) modules_install
+           $(MAKE) -C $(KDIR) M=$(PWD) INSTALL_MOD_PATH=$(DESTDIR) modules_install
            depmod -a

 clean:
```

With this I now get build errors:

```
# make -j16
make -C /lib/modules/5.14.0-gentoo/build M=/root/xt_tls/src modules
make[1]: Entering directory '/usr/src/linux-5.14.0-gentoo'
  CC [M]  /root/xt_tls/src/xt_tls_main.o
  CC [M]  /root/xt_tls/src/hostset.o
/root/xt_tls/src/hostset.c: In function ‘hs_init’:
/root/xt_tls/src/hostset.c:52:6: error: passing argument 4 of ‘proc_create_data’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   52 |      &proc_fops, hs);
      |      ^~~~~~~~~~
      |      |
      |      struct file_operations *
In file included from /root/xt_tls/src/hostset.c:9:
./include/linux/proc_fs.h:107:13: note: expected ‘const struct proc_ops *’ but argument is of type ‘struct file_operations *’
  107 |             const struct proc_ops *,
      |             ^~~~~~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[2]: *** [scripts/Makefile.build:271: /root/xt_tls/src/hostset.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:1851: /root/xt_tls/src] Error 2
make[1]: Leaving directory '/usr/src/linux-5.14.0-gentoo'
make: *** [Makefile:10: default] Error 2
```

Continuing not know what I'm actually doing: 
https://stackoverflow.com/a/61295957/2716218

```diff
diff --git a/src/hostset.c b/src/hostset.c
index 50b0fa0..e6b0b60 100644
--- a/src/hostset.c
+++ b/src/hostset.c
@@ -26,12 +26,11 @@ static ssize_t proc_write(struct file *file, const char __user *input,
 #define HIT_COUNT_DISPL_WIDTH 12
 #define HSE_SIZE_OVERHEAD (HIT_COUNT_DISPL_WIDTH + 2)

-static struct file_operations proc_fops = {
-    .owner = THIS_MODULE,
-    .open = seq_file_open,
-    .read = seq_read,
-    .write = proc_write,
-    .release = seq_release,
+static struct proc_ops proc_fops = {
+    .proc_open = seq_file_open,
+    .proc_read = seq_read,
+    .proc_write = proc_write,
+    .proc_release = seq_release,
 };

 // Initialize a host set
```

And it now builds, installs and modprobes

**Comment by caspChristian** [comment]
After running this for a while I see `seq_file: buggy .next function seq_read_next [xt_tls] did not update position index` from dmesg

**Comment by NiKiZe** [comment]
After running this for a while I see `seq_file: buggy .next function seq_read_next [xt_tls] did not update position index` from dmesg

**Comment by Lochnair** [comment]
Commited the fixes to have it working on 5.6+
Also hopefully fixed the warning about the position index
